### PR TITLE
ENCD-5926-adjust-biosample-characterizations-audit-to-check-parents

### PR DESCRIPTION
**Pull request by iyoungworth**
continuing-work

continuing-work

branch-switch

last-tests

### CODE REVIEW

File: src/encoded/audit/experiment.py [hunk 1]
```diff
@@ -1746,6 +1748,63 @@ def audit_experiment_biosample_characterization(value, system, excluded_types):
...
+                    and part_treatments == child_treatments
+                ):
+                    relationship = 'part_of'
+            elif (
+                biosample.get('originated_from')
+                and biosample.get('originated_from').get('characterizations')
+            ):
+                orig_ontology = biosample.get('originated_from').get('biosample_ontology')
+                orig_mods = sorted(biosample.get('originated_from').get('applied_modifications', []))
+                orig_treatments = sorted(biosample.get('originated_from').get('treatments', []))
```
Comment (gabdank) [inline src/encoded/audit/experiment.py]: I think this code could be rewritten to be something like:
part_of_biosample = biosample.get('part_of')
if (part_of_biosample and par_of_biosample.get('characterizations'):
    part_of_biosample.get('treatments')....
    ....
originated_from_biosample = biosample.get('originated_from')
if (originated_from_biosample and originated_From_biosample.get'characterizations'):
     originated_from_biosample.get('treatments').....
    ....

this way you are not "getting" the parent biosamples only once

File: src/encoded/audit/experiment.py [hunk 2]
```diff
@@ -3730,6 +3789,7 @@ def check_award_condition(experiment, awards):
         'replicates.library.biosample.applied_modifications',
         'replicates.library.biosample.applied_modifications.modified_site_by_target_id',
         'replicates.library.biosample.characterizations',
+        'replicates.library.biosample.part_of.characterizations',
```
Comment (gabdank) [inline src/encoded/audit/experiment.py:3795]: don't you need also "replicats.library.biosample.originated_from.characterizations' ?

File: src/encoded/tests/test_audit_experiment.py [hunk 3]
```diff
@@ -2194,6 +2194,139 @@ def test_audit_experiment_pooled_biosample_characterization(
...
+                       {'genetic_modifications': [interference_genetic_modification['@id']],
+                        'biosample_ontology': k562['uuid'],
+                        'donor': donor_1['@id'],
+                        'part_of': base_biosample['@id']})
+    testapp.patch_json(biosample_2['@id'],
+                       {'genetic_modifications': [interference_genetic_modification['@id']],
+                        'biosample_ontology': k562['uuid'],
+                        'donor': donor_1['@id'],
+                        'part_of': base_biosample['@id']})
+    testapp.patch_json(base_biosample['@id'],
```
Comment (gabdank) [inline src/encoded/tests/test_audit_experiment.py]: I guess same test needed for originated_from?

**Review by gabdank — DISMISSED** [review]

File: src/encoded/audit/experiment.py (hunk 2 above)
Comment (iyoungworth) [reply src/encoded/audit/experiment.py:3795]: Thank you for pointing this out, I've added that line

File: src/encoded/tests/test_audit_experiment.py (hunk 3 above)
Comment (iyoungworth) [reply src/encoded/tests/test_audit_experiment.py]: I trimmed down to the most useful tests and parametrized so they'd be run for both part_of and originated_from

File: src/encoded/audit/experiment.py (hunk 1 above)
Comment (iyoungworth) [reply src/encoded/audit/experiment.py]: Thank you for the suggestion, I applied this

File: src/encoded/audit/experiment.py [hunk 4]
```diff
@@ -1730,6 +1731,7 @@ def audit_experiment_biosample_characterization(value, system, excluded_types):
         ):
             continue
         biosample = rep['library']['biosample']
+        ontology = biosample['biosample_ontology']['@id']
```
Comment (hitz) [inline src/encoded/audit/experiment.py:1734]: this is actually a little dangerous unless you want a 500 error if the keys don't exist.  I think they are strictly require here so it's fine.

Comment (hitz) [reply src/encoded/audit/experiment.py:1734]: could wrap that in a try... except KeyError.

File: src/encoded/audit/experiment.py [hunk 5]
```diff
@@ -1746,6 +1748,65 @@ def audit_experiment_biosample_characterization(value, system, excluded_types):
...
+                if (
+                    orig_ontology == ontology
+                    and orig_mods == child_mods
+                    and orig_treatments == child_treatments
+                ):
+                    relationship = 'originated_from'
+            if relationship != '':
+                parent_characterization_status = {
+                    c.get('review', {}).get('status')
+                    for c in biosample.get(relationship).get('characterizations')
```
Comment (hitz) [inline src/encoded/audit/experiment.py:1788]: this is going to fail if biosample['relationship']['characterizations'] is not an iterable.  Maybe you secured that already?

**Review by hitz — COMMENTED** [review]
I am not sure if I need changes here or not @iyoungworth will slack you.

**Review by hitz — APPROVED** [review]

**Pull request merged**